# Pandas

In this notebook, we introduce you to Pandas which provides data structures and analysis tools for Data Science. Much of what we'll cover here involves the DataFrame data structure. As before, refer to documentation as and when you need to: https://pandas.pydata.org/

## Part 1

### Getting started with Pandas

Start by loading the `pandas` library (with alias `pd`).

In [1]:
# import the library
import pandas as pd


Load the provided dataset `data/airfoil.csv` using the [`pd.read_csv()`](https://pandas.pydata.org/pandas-docs/stable/generated/pandas.read_csv.html) function; call the DataFrame you load `df`.

In [2]:
# load the dataframe, use "head" to have a look
df = pd.read_csv("data/airfoil.csv")


Note that this `read_csv()` function is very flexible and can accomodate all sorts of file. 
For now, we're giving you a nicely formatted dataset that directly works well with Pandas.

`df` is a pandas DataFrame object. DataFrame objects have many methods and attributes which make data science easier! Use the `.head()` method to show what `df` looks like.

In [3]:
# Use .head() on df
df.head()


Frequency [Hz]  Angle [deg]  Chord length [m]  FS velocity [m/s]  \
0             800          0.0            0.3048               71.3   
1            1000          0.0            0.3048               71.3   
2            1250          0.0            0.3048               71.3   
3            1600          0.0            0.3048               71.3   
4            2000          0.0            0.3048               71.3   

   SSD thickness [m]  Sound pressure [dB]  
0           0.002663              126.201  
1           0.002663              125.201  
2           0.002663              125.951  
3           0.002663              127.591  
4           0.002663              127.461

Let's load another dataset into another dataframe and call it `smalldf`. The data is located at `data/smalldata.csv`. 

First, have a look at the data yourself by opening it with Excel or any text editor.

See what happens if you just run pd.read_csv() on it

In [4]:
# read it with pandas (don't save it, just allow it to print)
pd.read_csv('data/smalldata.csv')


F1   F2 F3      ID
0  15   70  M     Bob
1  18   50  F   Alice
2  15   55  F   Maria
3  12   85  M    John
4  14   68  M   Kevin
5  20  110  M  Donald

DataFrames have column names and an index. Import the csv and save it as a DataFrame called `smalldf`. Check the documentation (or google!) and:
* make the column names: `['Age', 'Weight', 'Gender', 'Name']`
* specify `name` as the index

Display `smalldf` in the notebook and check it against what you did above.

In [5]:
# add your code here to load smalldf
smalldf = pd.read_csv("data/smalldata.csv",
                      names=['Age', 'Weight', 'Gender', 'Name'],
                      index_col='Name', 
                      skiprows=1)
smalldf


Age  Weight Gender
Name                      
Bob      15      70      M
Alice    18      50      F
Maria    15      55      F
John     12      85      M
Kevin    14      68      M
Donald   20     110      M

### Retrieving some basic information

Now that you have a DataFrame object you can explore it (by typing `df.<TAB>` in a cell, you'll see all the methods and attributes)!

Examples of useful attributes:

* `shape` stores the dimensions of the data frame
* `columns` stores the names of the columns 
* `index` stores the names of the rows, by default pandas uses a range from 0 to the number of rows
* `dtypes` stores the `dtype` of each column

Show all of those, check it matches what you expected versus the output of `head` used earlier.

You can also use `df.describe()` to get a "description" of your data (per column, a number of standard statistics such as the mean, variance etc).

In [7]:
# add your code here to explore df's attributes
print(">> Shape attr:\n{}".format(df.shape))
print("\n>> Columns:\n{}".format(df.columns))
print("\n>> Index:\n{}".format(df.index))
print("\n>> Dtypes:\n{}".format(df.dtypes))


>> Shape attr:
(1503, 6)

>> Columns:
Index(['Frequency [Hz]', 'Angle [deg]', 'Chord length [m]',
       'FS velocity [m/s]', 'SSD thickness [m]', 'Sound pressure [dB]'],
      dtype='object')

>> Index:
RangeIndex(start=0, stop=1503, step=1)

>> Dtypes:
Frequency [Hz]           int64
Angle [deg]            float64
Chord length [m]       float64
FS velocity [m/s]      float64
SSD thickness [m]      float64
Sound pressure [dB]    float64
dtype: object


In [8]:
# use `df.describe()`
df.describe()


Frequency [Hz]  Angle [deg]  Chord length [m]  FS velocity [m/s]  \
count     1503.000000  1503.000000       1503.000000        1503.000000   
mean      2886.380572     6.782302          0.136548          50.860745   
std       3152.573137     5.918128          0.093541          15.572784   
min        200.000000     0.000000          0.025400          31.700000   
25%        800.000000     2.000000          0.050800          39.600000   
50%       1600.000000     5.400000          0.101600          39.600000   
75%       4000.000000     9.900000          0.228600          71.300000   
max      20000.000000    22.200000          0.304800          71.300000   

       SSD thickness [m]  Sound pressure [dB]  
count        1503.000000          1503.000000  
mean            0.011140           124.835943  
std             0.013150             6.898657  
min             0.000401           103.380000  
25%             0.002535           120.191000  
50%             0.004957           125.721000  
75%             0.015576           129.995500  
max             0.058411           140.987000

### Accessing elements in a dataframe

Let's get the value of the 1st column (`Age`) of the 3rd row (`Maria`) of `smalldf`. This can be done in many ways, the most convenient to you will depend on situation:

1. using `iloc`
1. using `loc`
1. each column of a `DataFrame` is a `Series` object: you can first get this then access relevant element

Try each of these methods below.

**Note**: remember that indexing in Python starts at 0.

In [9]:
# add your code here
print("1. {}".format(smalldf.iloc[2, 0]))
print("2. {}".format(smalldf.loc['Maria', 'Age']))
print("3. {}".format(smalldf['Age'][2]))


1. 15
2. 15
3. 15


### Using loc for fancy selections

Using `.loc`, can you retrieve the sub-dataframe of `df` with only the columns `SSD thickness [m]` and `Sound pressure [dB]`? Call this `df2`.

In [10]:
# add your code here
df2 = df.loc[:, ["SSD thickness [m]", "Sound pressure [dB]"]]
df2.head()


SSD thickness [m]  Sound pressure [dB]
0           0.002663              126.201
1           0.002663              125.201
2           0.002663              125.951
3           0.002663              127.591
4           0.002663              127.461

Using `to_csv`, output `df2` as a tab separated file (not comma) and call the file `airfoil_2.dat`.
(Open the file in an editor to check it matches what you expect).

In [12]:
df2.to_csv("data/airfoil_2.dat", sep='\t')

### Using masks to filter elements

Create two masks on our main dataframe `df`, one for all the frequencies and one for the sound pressure values. Using these masks get the dataframes with:

* Elements with frequencies higher than 1500 Hz
* Elements with sound pressure values lower than 126.0
* Elements where both of these conditions are true
* Elements where either of these conditions are true

In [13]:
# add your code here
freq_mask = df['Frequency [Hz]'] > 1500
sound_mask = df['Sound pressure [dB]'] < 126.0

print(df.loc[freq_mask].head())
print(df.loc[sound_mask].head())
print(df.loc[freq_mask & sound_mask].head())
print(df.loc[freq_mask | sound_mask].head())


   Frequency [Hz]  Angle [deg]  Chord length [m]  FS velocity [m/s]  \
3            1600          0.0            0.3048               71.3   
4            2000          0.0            0.3048               71.3   
5            2500          0.0            0.3048               71.3   
6            3150          0.0            0.3048               71.3   
7            4000          0.0            0.3048               71.3   

   SSD thickness [m]  Sound pressure [dB]  
3           0.002663              127.591  
4           0.002663              127.461  
5           0.002663              125.571  
6           0.002663              125.201  
7           0.002663              123.061  
   Frequency [Hz]  Angle [deg]  Chord length [m]  FS velocity [m/s]  \
1            1000          0.0            0.3048               71.3   
2            1250          0.0            0.3048               71.3   
5            2500          0.0            0.3048               71.3   
6            3150        

### Working with a pd.Series

Retrieve the series corresponding to the sound pressure from the dataframe, display

* show the name of the series (it should be `Sound pressure [dB]`)
* show the shape of the series (it should be `(1503,)`) 
* the mean and the median (resp. `124.84` and `125.72`)
* the mean of the squared values (it should be `15631.57`)
* Check [the documentation](https://pandas.pydata.org/pandas-docs/stable/generated/pandas.Series.html) and try some other methods and attributes

In [14]:
# add your code here
sp = df['Sound pressure [dB]']
print("Name: {}".format(sp.name))
print("Shape: {}".format(sp.shape))
print("Mean: {0:.2f}, Median: {1:.2f}".format(sp.mean(), sp.median()))
print("Mean of sq. vals: {0:.2f}".format((sp**2).mean()))


Name: Sound pressure [dB]
Shape: (1503,)
Mean: 124.84, Median: 125.72
Mean of sq. vals: 15631.57


### Getting the raw values

It can sometimes be useful to access the elements of a dataframe as a raw numpy array. For this you simply need to call the `.values` attribute on a series or attribute. 

* retrieve a raw array containing the sound pressures
* retrieve a raw array containing the frequencies and the sound pressures (watch the type inference!)

In [15]:
# add your code here
print(df['Sound pressure [dB]'].values)

print(df[['Frequency [Hz]', 'Sound pressure [dB]']].values)


[126.201 125.201 125.951 ... 106.604 106.224 104.204]
[[ 800.     126.201]
 [1000.     125.201]
 [1250.     125.951]
 ...
 [4000.     106.604]
 [5000.     106.224]
 [6300.     104.204]]


## Part 2

## More useful tricks with Pandas

Join the column `Frequency [Hz]` and `df2` using `concat`, call the result `df3` and display the summary statistics.

In [16]:
# add your code here to create df3
df3 = pd.concat([df['Frequency [Hz]'], df2], axis=1)
df3.describe([0.1, 0.9])


Frequency [Hz]  SSD thickness [m]  Sound pressure [dB]
count     1503.000000        1503.000000          1503.000000
mean      2886.380572           0.011140           124.835943
std       3152.573137           0.013150             6.898657
min        200.000000           0.000401           103.380000
10%        400.000000           0.001211           115.299200
50%       1600.000000           0.004957           125.721000
90%       6300.000000           0.033779           133.234000
max      20000.000000           0.058411           140.987000

### Groupby and apply

Group the rows of `df3` by Frequency and call the result `freq_df3`, show the length of this object, does it correspond to what you expected?

Then:
* Using directly `.mean()` on your groupby object, compute the mean of each features per group of frequencies.
* Using `apply`, compute the max of each features per group of frequencies (passing the `max` function to `apply`)
* Using `agg`, compute the mean and the max of the Chord length and the FS velocity

Tip: `apply` and `agg` expect you to pass the functions you want to apply on the groupby object. Instead of implementing those functions (mean and max) yourself, you can import the library `numpy` that has them already available.

In [17]:
# add your code here to compute freq_df3
freq_df3 = df3.groupby(by='Frequency [Hz]')
len(freq_df3)


21

In [18]:
# add your code here to compute the means
import numpy as np

freq_df3.mean()
freq_df3.apply(np.max)
freq_df3.agg([np.max, np.min, np.sum, np.mean])


SSD thickness [m]                                \
                            amax      amin       sum      mean   
Frequency [Hz]                                                   
200                     0.058411  0.003101  0.831930  0.023769   
250                     0.058411  0.003101  0.888236  0.021148   
315                     0.058411  0.002293  0.966862  0.017265   
400                     0.058411  0.002293  1.026562  0.014878   
500                     0.058411  0.000907  1.049153  0.013451   
630                     0.058411  0.000907  1.073252  0.012196   
800                     0.058411  0.000740  1.091312  0.011251   
1000                    0.058411  0.000439  1.093278  0.011043   
1250                    0.058411  0.000439  1.095421  0.010954   
1600                    0.058411  0.000428  1.098151  0.010662   
2000                    0.058411  0.000412  1.099412  0.010471   
2500                    0.058411  0.000401  1.088103  0.010463   
3150                    0.058411  0.000401  1.082889  0.010513   
4000                    0.058411  0.000401  1.078106  0.010570   
5000                    0.058411  0.000401  0.923188  0.009718   
6300                    0.058411  0.000401  0.848861  0.009538   
8000                    0.026427  0.000401  0.210053  0.004039   
10000                   0.013025  0.000401  0.122250  0.002911   
12500                   0.004978  0.000401  0.046191  0.001848   
16000                   0.004006  0.000401  0.021347  0.001642   
20000                   0.004006  0.000401  0.008684  0.001447   

               Sound pressure [dB]                                  
                              amax     amin        sum        mean  
Frequency [Hz]                                                      
200                        138.758  109.951   4357.891  124.511171  
250                        139.918  110.491   5262.622  125.300524  
315                        139.808  111.911   7039.890  125.712321  
400                        139.438  115.444   8722.363  126.411058  
500                        140.158  114.900   9880.486  126.672897  
630                        138.423  116.550  11258.035  127.932216  
800                        139.226  116.560  12485.019  128.711536  
1000                       138.274  114.670  12729.823  128.584071  
1250                       140.987  112.160  12814.387  128.143870  
1600                       136.678  110.780  13061.985  126.815388  
2000                       134.976  109.520  13199.946  125.713771  
2500                       135.956  106.880  12995.092  124.952808  
3150                       137.007  106.260  12761.019  123.893388  
4000                       138.557  104.500  12469.921  122.254127  
5000                       138.607  104.130  11504.592  121.100968  
6300                       138.537  103.380  10619.665  119.322079  
8000                       137.207  110.515   6270.392  120.584462  
10000                      134.563  108.265   4980.925  118.593452  
12500                      133.207  106.111   2969.221  118.768840  
16000                      130.477  106.582   1534.894  118.068769  
20000                      123.217  109.638    710.254  118.375667

### (Bonus) Joining dataframes

Check the pandas documentation on the [.join method](https://pandas.pydata.org/pandas-docs/stable/generated/pandas.DataFrame.join.html) to work out how to join `smalldf` with the following data:

```
{'Salary': [100, 150, 110, 90, 105, 500], 'Education': [5, 10, 7, 3, 4, 0]}
```

In [19]:
# add your code here
# the index *must* be specified
otherdf = pd.DataFrame({'Salary': [100, 150, 110, 90, 105, 500], 
                        'Education': [5, 10, 7, 3, 4, 0]}, 
                        index=smalldf.index)

completedf = smalldf.join(otherdf)
completedf.head()


Age  Weight Gender  Salary  Education
Name                                        
Bob     15      70      M     100          5
Alice   18      50      F     150         10
Maria   15      55      F     110          7
John    12      85      M      90          3
Kevin   14      68      M     105          4